android/java/org/chromium/chrome/browser/crypto_wallet/fragments/dapps/ConnectAccountFragment.java

```diff
@@ -0,0 +1,256 @@
+            updateAccounts();
+        }
+    }
+
+    @Override
+    public void onDestroy() {
+        super.onDestroy();
+        if (mFaviconHelper != null) mFaviconHelper.destroy();
+        if (mDefaultFaviconHelper != null) mDefaultFaviconHelper.clearCache();
+    }
+
+    @Override
+    public HashSet<AccountInfo> getAccountsWithPermissions() {
+        return mAccountsWithPermissions;
+    }
+
+    @Override
+    public String getSelectedAccount() {
+        return mSelectedAccount;
+    }
+
+    @Override
+    public void connectAccount(AccountInfo account) {
+        getBraveWalletService().addEthereumPermission(
+                Utils.originFromGURL(getCurrentHostHttpAddress()), account.address, success -> {
+                    if (!success) {
+                        return;
+                    }
+                    getKeyringService().setSelectedAccount(
+                            account.address, CoinType.ETH, setSuccess -> {
+                                if (setSuccess) {
+                                    updateAccounts();
+                                }
+                            });
+                });
+    }
+
+    @Override
+    public void disconnectAccount(AccountInfo account) {
+        getBraveWalletService().resetEthereumPermission(
+                Utils.originFromGURL(getCurrentHostHttpAddress()), account.address, success -> {
+                    if (!success) {
+                        return;
+                    }
+                    if (!mSelectedAccount.equals(account.address)) {
+                        updateAccounts();
+
+                        return;
+                    }
+
+                    boolean updateCalled = false;
+                    assert mAccountsWithPermissions != null;
+                    Iterator<AccountInfo> it = mAccountsWithPermissions.iterator();
+                    while (it.hasNext()) {
+                        String currentAddress = it.next().address;
+                        if (!currentAddress.equals(account.address)) {
+                            updateCalled = true;
+                            getKeyringService().setSelectedAccount(currentAddress, CoinType.ETH,
+                                    setSuccess -> { updateAccounts(); });
+                            break;
+                        }
+                    }
+                    if (!updateCalled) {
+                        updateAccounts();
+                    }
+                });
+    }
+
+    @Override
+    public void switchAccount(AccountInfo account) {
+        getKeyringService().setSelectedAccount(account.address, CoinType.ETH, setSuccess -> {
+            if (setSuccess) {
+                updateAccounts();
+            }
+        });
+    }
+
+    private GURL getCurrentHostHttpAddress() {
+        ChromeTabbedActivity activity = BraveActivity.getChromeTabbedActivity();
+        if (activity != null) {
+            return activity.getActivityTab().getUrl().getOrigin();
```

Comment:
It appears `getUrl()` is not a last committed url and might be unsafe to use it.
Maybe smth like `activity.getActivityTab().getWebContents().getMainFrame().getLastCommittedOrigin()` will give required origin and we don't need `Utils.originFromGURL()` at all

Reply:
the question is only is how to convert from Origin to GURL. I cannot find anything related to that. We use the above function to display current URL as well, so I need some way of conversion. For mojo call yes, I can use Origin directly.

Reply:
done